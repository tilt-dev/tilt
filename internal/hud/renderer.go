package hud

import (
	"fmt"
	"os"
	"strings"
	"sync"
	"time"

	"github.com/gdamore/tcell"
	"github.com/windmilleng/tilt/internal/dockercompose"
	"github.com/windmilleng/tilt/internal/hud/view"
	"github.com/windmilleng/tilt/internal/model"
	"github.com/windmilleng/tilt/internal/rty"
)

type Renderer struct {
	rty    rty.RTY
	screen tcell.Screen
	mu     *sync.Mutex
	clock  func() time.Time
}

func NewRenderer(clock func() time.Time) *Renderer {
	return &Renderer{
		mu:    new(sync.Mutex),
		clock: clock,
	}
}

func (r *Renderer) Render(v view.View, vs view.ViewState) error {
	r.mu.Lock()
	defer r.mu.Unlock()
	if r.rty != nil {
		layout := r.layout(v, vs)
		err := r.rty.Render(layout)
		if err != nil {
			return err
		}
	}
	return nil
}

var cText = tcell.Color232
var cLightText = tcell.Color243
var cGood = tcell.ColorGreen
var cBad = tcell.ColorRed
var cPending = tcell.ColorYellow

var statusColors = map[string]tcell.Color{
	"Running":                          cGood,
	"ContainerCreating":                cPending,
	"Pending":                          cPending,
	"Error":                            cBad,
	"CrashLoopBackOff":                 cBad,
	string(dockercompose.StatusInProg): cPending,
	string(dockercompose.StatusUp):     cGood,
	string(dockercompose.StatusDown):   cBad,
	"Completed":                        cGood,
}

func (r *Renderer) layout(v view.View, vs view.ViewState) rty.Component {
	l := rty.NewFlexLayout(rty.DirVert)
	if vs.ShowNarration {
		l.Add(renderNarration(vs.NarrationMessage))
		l.Add(rty.NewLine())
	}

	l.Add(r.renderResourceHeader(v))
	l.Add(r.renderResources(v, vs))
	l.Add(r.renderPaneHeader(vs))
	l.Add(r.renderLogPane(v, vs))
	l.Add(r.renderFooter(v, keyLegend(v, vs)))

	var ret rty.Component = l
	if vs.LogModal.TiltLog == view.TiltLogFullScreen {
		ret = r.renderTiltLog(v, vs, keyLegend(v, vs), ret)
	} else if vs.LogModal.ResourceLogNumber != 0 {
		ret = r.renderResourceLogModal(v.Resources[vs.LogModal.ResourceLogNumber-1], ret)
	}

	ret = r.maybeAddAlertModal(vs, ret)

	return ret
}

func (r *Renderer) maybeAddAlertModal(vs view.ViewState, layout rty.Component) rty.Component {
	if vs.AlertMessage != "" {
		l := rty.NewLines()
		l.Add(rty.TextString(""))

		msg := "   " + vs.AlertMessage + "   "
		l.Add(rty.Fg(rty.TextString(msg), tcell.ColorDefault))
		l.Add(rty.TextString(""))

		b := rty.NewBox(l)
		b.SetTitle("! Alert !")
		layout = r.renderModal(rty.Fg(b, tcell.ColorRed), layout, false)
	}
	return layout
}

func (r *Renderer) renderLogPane(v view.View, vs view.ViewState) rty.Component {
	l := rty.NewConcatLayout(rty.DirHor)
	log := rty.NewTextScrollLayout("log")
	log.Add(rty.TextString(v.Log))
	l.Add(log)
	height := 7
	if vs.LogModal.TiltLog == view.TiltLogMinimized {
		height = 1
	}
	return rty.NewFixedSize(l, rty.GROW, height)
}

func (r *Renderer) renderPaneHeader(vs view.ViewState) rty.Component {
	var s string
	switch vs.LogModal.TiltLog {
	case view.TiltLogFullScreen:
		s = "(l) minimize log ↓"
	case view.TiltLogPane:
		s = "(l) maximize log ↑"
	case view.TiltLogMinimized:
		s = "(l) expand log ↑"
	}
	l := rty.NewLine()
	l.Add(rty.NewFillerString('─'))
	l.Add(rty.TextString(fmt.Sprintf(" %s ", s)))
	l.Add(rty.NewFillerString('─'))
	return l
}

func (r *Renderer) renderStatusBar(v view.View) rty.Component {
	sb := rty.NewStringBuilder()
	sb.Text(" ") // Indent
	errorCount := 0
	for _, res := range v.Resources {
		if isInError(res, v.TriggerMode) {
			errorCount++
		}
	}
	if errorCount == 0 && v.TiltfileErrorMessage == "" {
		sb.Fg(cGood).Text("✓").Fg(tcell.ColorDefault).Fg(cText).Text(" OK").Fg(tcell.ColorDefault)
	} else {
		var errorCountMessage string
		var tiltfileError strings.Builder
		s := "error"
		if errorCount > 1 {
			s = "errors"
		}

		if errorCount > 0 {
			errorCountMessage = fmt.Sprintf(" %d %s", errorCount, s)
		}

		if v.TiltfileErrorMessage != "" {
			_, err := tiltfileError.WriteString(" • Tiltfile error")
			if err != nil {
				// This space intentionally left blank
			}
		}
		sb.Fg(cBad).Text("✖").Fg(tcell.ColorDefault).Fg(cText).Textf("%s%s", errorCountMessage, tiltfileError.String()).Fg(tcell.ColorDefault)
	}
	return rty.Bg(rty.OneLine(sb.Build()), tcell.ColorWhiteSmoke)
}

func (r *Renderer) renderFooter(v view.View, keys string) rty.Component {
	footer := rty.NewConcatLayout(rty.DirVert)
	footer.Add(r.renderStatusBar(v))
	l := rty.NewConcatLayout(rty.DirHor)
	sbRight := rty.NewStringBuilder()
	sbRight.Text(keys)
	l.AddDynamic(rty.NewFillerString(' '))
	l.Add(sbRight.Build())
	footer.Add(l)
	return rty.NewFixedSize(footer, rty.GROW, 2)
}

func keyLegend(v view.View, vs view.ViewState) string {
	defaultKeys := "Browse (↓ ↑), Expand (→) ┊ (enter) log, (b)rowser ┊ (ctrl-C) quit  "
	if vs.LogModal.TiltLog == view.TiltLogFullScreen {
		return "Scroll (↓ ↑) ┊ cycle (l)og view "
	} else if vs.LogModal.ResourceLogNumber != 0 {
		return "Scroll (↓ ↑) ┊ (esc) close logs "
	} else if vs.AlertMessage != "" {
		return "Tilt (l)og ┊ (esc) close alert "
	} else if v.TriggerMode == model.TriggerManual {
		return "Build (space) ┊ " + defaultKeys
	}
	return defaultKeys
}

func isInError(res view.Resource, triggerMode model.TriggerMode) bool {
	return statusColor(res, triggerMode) == cBad
}

func isCrashing(res view.Resource) bool {
	return (res.IsK8S() && res.K8SInfo().PodRestarts > 0) ||
		res.LastBuild().Reason.Has(model.BuildReasonFlagCrash) ||
		res.CurrentBuild.Reason.Has(model.BuildReasonFlagCrash) ||
		res.PendingBuildReason.Has(model.BuildReasonFlagCrash) ||
		res.IsDC() && res.DockerComposeTarget().Status() == string(dockercompose.StatusCrash)
}

func bestLogs(res view.Resource) string {
	// A build is in progress, triggered by an explicit edit.
	if res.CurrentBuild.StartTime.After(res.LastBuild().FinishTime) &&
		!res.CurrentBuild.Reason.IsCrashOnly() {
		return string(res.CurrentBuild.Log)
	}

	// A build is in progress, triggered by a pod crash.
	if res.CurrentBuild.StartTime.After(res.LastBuild().FinishTime) &&
		res.CurrentBuild.Reason.IsCrashOnly() {
		return res.CrashLog + "\n\n" + string(res.CurrentBuild.Log)
	}

	// The last build was an error.
	if res.LastBuild().Error != nil {
		return string(res.LastBuild().Log)
	}

	if k8sInfo, ok := res.ResourceInfo.(view.K8SResourceInfo); ok {
		// Two cases:
		// 1) The last build finished before this pod started
		// 2) This log is from an in-place container update.
		// in either case, prepend them to pod logs.
		if (res.LastBuild().StartTime.Equal(k8sInfo.PodUpdateStartTime) ||
			res.LastBuild().StartTime.Before(k8sInfo.PodCreationTime)) &&
			len(res.LastBuild().Log) > 0 {
			return string(res.LastBuild().Log) + "\n" + res.ResourceInfo.RuntimeLog()
		}

		// The last build finished, but the pod hasn't started yet.
		if res.LastBuild().StartTime.After(k8sInfo.PodCreationTime) {
			return string(res.LastBuild().Log)
		}
	}

	if res.IsTiltfile {
		return string(res.LastBuild().Log)
	}

	return string(res.LastBuild().Log) + "\n" + res.ResourceInfo.RuntimeLog()
}

func (r *Renderer) renderTiltLog(v view.View, vs view.ViewState, keys string, background rty.Component) rty.Component {
	l := rty.NewConcatLayout(rty.DirVert)
	sl := rty.NewTextScrollLayout(logScrollerName)
	l.Add(r.renderPaneHeader(vs))
	sl.Add(rty.TextString(v.Log))
	l.AddDynamic(sl)
	l.Add(r.renderFooter(v, keys))
	return rty.NewModalLayout(background, l, 1, true)
}

func (r *Renderer) renderResourceLogModal(res view.Resource, background rty.Component) rty.Component {
	s := bestLogs(res)
	if len(strings.TrimSpace(s)) == 0 {
		s = fmt.Sprintf("No log output for %s", res.Name)
	}

	l := rty.NewTextScrollLayout(logScrollerName)
	l.Add(rty.TextString(s))
	box := rty.NewGrowingBox()
	box.SetInner(l)
	box.SetTitle(fmt.Sprintf("LOG: %s", res.Name))

	return rty.NewModalLayout(background, box, 0.9, true)
}

func (r *Renderer) renderModal(fg rty.Component, bg rty.Component, fixed bool) rty.Component {
	return rty.NewModalLayout(bg, fg, .9, fixed)
}

func renderNarration(msg string) rty.Component {
	lines := rty.NewLines()
	l := rty.NewLine()
	l.Add(rty.TextString(msg))
	lines.Add(rty.NewLine())
	lines.Add(l)
	lines.Add(rty.NewLine())

	box := rty.Fg(rty.Bg(lines, tcell.ColorLightGrey), cText)
	return rty.NewFixedSize(box, rty.GROW, 3)
}

func (r *Renderer) renderResourceHeader(v view.View) rty.Component {
	l := rty.NewConcatLayout(rty.DirHor)
	l.Add(rty.ColoredString("  RESOURCE NAME ", cLightText))
	l.AddDynamic(rty.NewFillerString(' '))

	k8sCell := rty.ColoredString(" DEPLOY", cLightText)
	l.Add(k8sCell)
	l.Add(middotText())

	buildCell := rty.NewMinLengthLayout(BuildDurCellMinWidth+BuildStatusCellMinWidth, rty.DirHor).
		SetAlign(rty.AlignEnd).
		Add(rty.ColoredString("BUILD STATUS", cLightText))
	l.Add(buildCell)
	l.Add(middotText())
	deployCell := rty.NewMinLengthLayout(DeployCellMinWidth+1, rty.DirHor).
		SetAlign(rty.AlignEnd).
		Add(rty.ColoredString("UPDATED ", cLightText))
	l.Add(deployCell)
	return rty.OneLine(l)
}

func (r *Renderer) renderResources(v view.View, vs view.ViewState) rty.Component {
	rs := v.Resources

	cl := rty.NewConcatLayout(rty.DirVert)

	childNames := make([]string, len(rs))
	for i, r := range rs {
		childNames[i] = r.Name.String()
	}
	// the items added to `l` below must be kept in sync with `childNames` above
	l, selectedResource := r.rty.RegisterElementScroll(resourcesScollerName, childNames)

	if len(rs) > 0 {
		for i, res := range rs {
			l.Add(r.renderResource(res, vs.Resources[i], v.TriggerMode, selectedResource == res.Name.String()))
		}
	}

	cl.Add(l)
	return cl
}

func (r *Renderer) renderResource(res view.Resource, rv view.ResourceViewState, triggerMode model.TriggerMode, selected bool) rty.Component {
	return NewResourceView(res, rv, triggerMode, selected, r.clock).Build()
}

func (r *Renderer) SetUp() (chan tcell.Event, error) {
	r.mu.Lock()
	defer r.mu.Unlock()

	screen, err := tcell.NewScreen()
	if err != nil {
		if err == tcell.ErrTermNotFound {
			// The statically-compiled tcell only supports the most common TERM configs.
			// The dynamically-compiled tcell supports more, but has distribution problems.
			// See: https://github.com/gdamore/tcell/issues/252
			term := os.Getenv("TERM")
			return nil, fmt.Errorf("Tilt does not support TERM=%q. "+
				"This is not a common Terminal config. "+
				"If you expect that you're using a common terminal, "+
				"you might have misconfigured $TERM in your .profile.", term)
		}
		return nil, err
	}
	if err = screen.Init(); err != nil {
		return nil, err
	}
	screenEvents := make(chan tcell.Event)
	go func() {
		for {
			screenEvents <- screen.PollEvent()
		}
	}()

	r.rty = rty.NewRTY(screen)

	r.screen = screen

	return screenEvents, nil
}

func (r *Renderer) Reset() {
	r.mu.Lock()
	defer r.mu.Unlock()

	if r.screen != nil {
		r.screen.Fini()
	}

	r.screen = nil
}
