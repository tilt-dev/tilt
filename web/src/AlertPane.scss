@import "constants";

.AlertPane {
  padding: $spacing-unit / 2;
  width: 100%;
  display: flex;
  justify-content: center;
  box-sizing: border-box;
}

.AlertPane ul {
  list-style: none;
  width: 100%;
}
.AlertPane-item {
  position: relative;
}
.AlertPane-item > header {
  background-color: $color-gray;
  border: 1px solid $color-gray-light;
  display: flex;
  flex-direction: column;
  padding-top: $spacing-unit / 4;
  padding-bottom: $spacing-unit / 4;
  padding-left: $spacing-unit / 2;
  padding-right: $spacing-unit / 2;
}

.AlertPane-headerDiv {
  display: flex;
  justify-content: space-between;
}

.AlertPane-headerDiv span + span {
  margin-left: $spacing-unit / 2;
  border-left: 1px solid $color-gray-light;
  padding-left: $spacing-unit / 2;
}

.AlertPane-headerDiv-header {
  margin: 0;
  font-size: $font-size;
}

.AlertPane-item > section {
  background-color: $color-gray-dark;
  border-bottom-left-radius: $spacing-unit / 4;
  border-bottom-right-radius: $spacing-unit / 4;
  border-left: 1px solid $color-gray-light;
  border-right: 1px solid $color-gray-light;
  border-bottom: 1px solid $color-gray-light;
  padding: $spacing-unit / 2;
  min-height: 2em;
}

.AlertPane-item + .AlertPane-item {
  margin-top: $spacing-unit;
}

// Hide all lines by default
.AlertPane-item code {
  display: none;
  white-space: pre-wrap;
}

// ...only show last few lines
.AlertPane-item code:nth-last-child(-n + 8) {
  display: inline;
  min-height: $spacing-unit / 2; // Respect blank lines
  height: auto;
  opacity: 1;
}

.AlertPane-dismissButton {
  position: absolute;
  right: $spacing-unit/2;
  bottom: $spacing-unit/2;
  align-content: center;
  background-color: $color-gray-dark;
  color: $color-blue;
  @include button-text;
  border: 1px solid $color-blue;
  border-radius: 5px;
  padding: $spacing-unit/3;
  min-width: $spacing-unit * 3;
  transition: background-color ease 0.3s;
}

.AlertPane-dismissButton:hover {
  background-color: $color-gray;
  cursor: pointer;
}