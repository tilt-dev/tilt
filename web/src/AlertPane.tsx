import React, { PureComponent } from "react"
import { ReactComponent as LogoWordmarkSvg } from "./assets/svg/logo-wordmark-gray.svg"
import AnsiLine from "./AnsiLine"
import TimeAgo from "react-timeago"
import "./AlertPane.scss"
import { timeAgoFormatter } from "./timeFormatters"
import { getResourceAlerts, hasAlert } from "./alerts"
import PathBuilder from "./PathBuilder"
import LogStore from "./LogStore"

type Resource = Proto.webviewResource

type AlertsProps = {
  pathBuilder: PathBuilder
  resources: Array<Resource>
  logStore: LogStore | null
}

function logToLines(s: string) {
  return s.split("\n").map((l, i) => <AnsiLine key={"logLine" + i} line={l} />)
}

class AlertPane extends PureComponent<AlertsProps> {
  render() {
    let el = (
      <section className="Pane-empty-message">
        <LogoWordmarkSvg />
        <h2>No Alerts Found</h2>
      </section>
    )

    let alerts = this.renderAlerts()
    if (alerts.length > 0) {
      el = <ul>{alerts}</ul>
    }

    return <section className="AlertPane">{el}</section>
  }

  renderAlerts() {
    let formatter = timeAgoFormatter
    let alertElements: Array<JSX.Element> = []
    let resources = this.props.resources
    let isSnapshot = this.props.pathBuilder.isSnapshot()

    let alertResources = resources.filter(r => hasAlert(r))
    alertResources.forEach(resource => {
      let resName = resource.name ?? ""
      getResourceAlerts(resource, this.props.logStore).forEach(alert => {
        let dismissButton = <div />
        if (alert.dismissHandler && !isSnapshot) {
          dismissButton = (
            <button
              className="AlertPane-dismissButton"
              onClick={alert.dismissHandler}
            >
              Dismiss
            </button>
          )
        }
        alertElements.push(
          <li key={alert.alertType + resName} className="AlertPane-item">
            <header>
              <div className="AlertPane-headerDiv">
                <h3 className="AlertPane-headerDiv-header">{alert.header}</h3>
              </div>
              <div className="AlertPane-headerDiv">
                <p>
                  <span>Resource: {alert.resourceName}</span>
                  <span>Type: {alert.alertType}</span>
                </p>
                <TimeAgo date={alert.timestamp} formatter={formatter} />
              </div>
            </header>
            <section>{logToLines(alert.msg)}</section>
            {dismissButton}
          </li>
        )
      })
    })
    return alertElements
  }
}

export default AlertPane
