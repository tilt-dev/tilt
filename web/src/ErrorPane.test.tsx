import React from "react"
import AlertPane, { AlertResource } from "./ErrorPane"
import renderer from "react-test-renderer"

beforeEach(() => {
  Date.now = jest.fn(() => 1482363367071)
})

it("renders no errors", () => {
  let resources = [
    {
      Name: "foo",
      BuildHistory: [],
      ResourceInfo: {},
    },
  ]

  const tree = renderer
    .create(<AlertPane resources={resources.map(r => new AlertResource(r))} />)
    .toJSON()

  expect(tree).toMatchSnapshot()
})

it("renders one build error", () => {
  const ts = "1,555,970,585,039"
  let resources = [
    {
      Name: "foo",
      BuildHistory: [
        {
          Log: "laa dee daa I'm an error\nfor real I am",
          FinishTime: ts,
          Error: {},
        },
      ],
      ResourceInfo: {},
    },
  ]

  const tree = renderer
    .create(<AlertPane resources={resources.map(r => new AlertResource(r))} />)
    .toJSON()

  expect(tree).toMatchSnapshot()
})

it("renders the last build with an error", () => {
  const ts = "1,555,970,585,039"
  let resources = [
    {
      Name: "foo",
      BuildHistory: [
        {
          Log: "laa dee daa I'm another error\nBetter watch out",
          FinishTime: ts,
          Error: {},
        },
        {
          Log: "laa dee daa I'm an error\nI'm serious",
          FinishTime: ts,
          Error: {},
        },
      ],
      ResourceInfo: {},
    },
  ]

  const tree = renderer
    .create(<AlertPane resources={resources.map(r => new AlertResource(r))} />)
    .toJSON()

  expect(tree).toMatchSnapshot()
})

it("renders one container start error", () => {
  const ts = "1,555,970,585,039"
  let resources = [
    {
      Name: "foo",
      CrashLog: "Eeeeek there is a problem",
      BuildHistory: [
        {
          Log: "laa dee daa I'm an error\nI'm serious",
          FinishTime: ts,
          Error: null,
        },
      ],
      ResourceInfo: {
        PodCreationTime: ts,
        PodStatus: "Error",
        PodRestarts: 2,
      },
    },
  ]

  const tree = renderer
    .create(<AlertPane resources={resources.map(r => new AlertResource(r))} />)
    .toJSON()
  expect(tree).toMatchSnapshot()

  // the podStatus will flap between "Error" and "CrashLoopBackoff"
  resources = [
    {
      Name: "foo",
      CrashLog: "Eeeeek there is a problem",
      BuildHistory: [
        {
          Log: "laa dee daa I'm not an error\nI'm serious",
          FinishTime: ts,
          Error: null,
        },
      ],
      ResourceInfo: {
        PodCreationTime: ts,
        PodStatus: "CrashLoopBackOff",
        PodRestarts: 3,
      },
    },
  ]

  const newTree = renderer
    .create(<AlertPane resources={resources.map(r => new AlertResource(r))} />)
    .toJSON()
  expect(newTree).toMatchSnapshot()
})

it("shows that a container has restarted", () => {
  const ts = "1,555,970,585,039"
  const resources = [
    {
      Name: "foo",
      CrashLog: "Eeeeek the container crashed",
      BuildHistory: [
        {
          Log: "laa dee daa I'm not an error\nseriously",
          FinishTime: ts,
          Error: null,
        },
      ],
      ResourceInfo: {
        PodCreationTime: ts,
        PodStatus: "ok",
        PodRestarts: 1,
      },
    },
  ]

  const tree = renderer
    .create(<AlertPane resources={resources.map(r => new AlertResource(r))} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it("shows that a crash rebuild has occurred", () => {
  const ts = "1,555,970,585,039"
  const resources = [
    {
      Name: "foo",
      CrashLog: "Eeeeek the container crashed",
      BuildHistory: [
        {
          Log: "laa dee daa I'm not an error\nseriously",
          FinishTime: ts,
          Error: null,
          IsCrashRebuild: true,
        },
      ],
      ResourceInfo: {
        PodCreationTime: ts,
        PodStatus: "ok",
      },
    },
  ]

  const tree = renderer
    .create(<AlertPane resources={resources.map(r => new AlertResource(r))} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it("renders multiple lines of a crash log", () => {
  const ts = "1,555,970,585,039"
  const resources = [
    {
      Name: "foo",
      CrashLog: "Eeeeek the container crashed\nno but really it crashed",
      BuildHistory: [
        {
          Log: "laa dee daa I'm not an error\nseriously",
          FinishTime: ts,
          Error: null,
          IsCrashRebuild: true,
        },
      ],
      ResourceInfo: {
        PodCreationTime: ts,
        PodStatus: "ok",
      },
    },
  ]

  const tree = renderer
    .create(<AlertPane resources={resources.map(r => new AlertResource(r))} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})
