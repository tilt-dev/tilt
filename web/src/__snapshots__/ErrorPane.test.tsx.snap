// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders multiple lines of a crash log 1`] = `
<section
  className="ErrorPane"
>
  <ul>
    <li
      className="ErrorPane-item"
    >
      <header>
        <p>
          foo
        </p>
        <p>
          Pod crashed!
        </p>
      </header>
      <section>
        <code>
          <span
            className={null}
            style={null}
          >
            Eeeeek the container crashed
          </span>
        </code>
        <code>
          <span
            className={null}
            style={null}
          >
            no but really it crashed
          </span>
        </code>
      </section>
    </li>
  </ul>
</section>
`;

exports[`renders no errors 1`] = `
<section
  className="ErrorPane"
>
  <section
    className="Pane-empty-message"
  >
    <svg>
      logo-wordmark-gray.svg
    </svg>
    <h2>
      No Alerts Found
    </h2>
  </section>
</section>
`;

exports[`renders one build error 1`] = `
<section
  className="ErrorPane"
>
  <ul>
    <li
      className="ErrorPane-item"
    >
      <header>
        <p>
          foo
        </p>
        <p>
          <time
            dateTime="1949-11-18T09:39:00.000Z"
            title="1,555,970,585,039"
          >
            67yr
          </time>
        </p>
      </header>
      <section>
        <code>
          <span
            className={null}
            style={null}
          >
            laa dee daa I'm an error
          </span>
        </code>
        <code>
          <span
            className={null}
            style={null}
          >
            for real I am
          </span>
        </code>
      </section>
    </li>
  </ul>
</section>
`;

exports[`renders one container start error 1`] = `
<section
  className="ErrorPane"
>
  <ul>
    <li
      className="ErrorPane-item"
    >
      <header>
        <p>
          foo
        </p>
        <p>
          <time
            dateTime="1949-11-18T09:39:00.000Z"
            title="1,555,970,585,039"
          >
            67yr
          </time>
        </p>
      </header>
      <section>
        <code>
          <span
            className={null}
            style={null}
          >
            Eeeeek there is a problem
          </span>
        </code>
      </section>
    </li>
  </ul>
</section>
`;

exports[`renders one container start error 2`] = `
<section
  className="ErrorPane"
>
  <ul>
    <li
      className="ErrorPane-item"
    >
      <header>
        <p>
          foo
        </p>
        <p>
          <time
            dateTime="1949-11-18T09:39:00.000Z"
            title="1,555,970,585,039"
          >
            67yr
          </time>
        </p>
      </header>
      <section>
        <code>
          <span
            className={null}
            style={null}
          >
            Eeeeek there is a problem
          </span>
        </code>
      </section>
    </li>
  </ul>
</section>
`;

exports[`renders the last build with an error 1`] = `
<section
  className="ErrorPane"
>
  <ul>
    <li
      className="ErrorPane-item"
    >
      <header>
        <p>
          foo
        </p>
        <p>
          <time
            dateTime="1949-11-18T09:39:00.000Z"
            title="1,555,970,585,039"
          >
            67yr
          </time>
        </p>
      </header>
      <section>
        <code>
          <span
            className={null}
            style={null}
          >
            laa dee daa I'm another error
          </span>
        </code>
        <code>
          <span
            className={null}
            style={null}
          >
            Better watch out
          </span>
        </code>
      </section>
    </li>
  </ul>
</section>
`;

exports[`shows that a container has restarted 1`] = `
<section
  className="ErrorPane"
>
  <ul>
    <li
      className="ErrorPane-item"
    >
      <header>
        <p>
          foo
        </p>
        <p>
          Restarts: 1
        </p>
      </header>
      <section>
        <code>
          <span
            className={null}
            style={null}
          >
            Eeeeek the container crashed
          </span>
        </code>
      </section>
    </li>
  </ul>
</section>
`;

exports[`shows that a crash rebuild has occurred 1`] = `
<section
  className="ErrorPane"
>
  <ul>
    <li
      className="ErrorPane-item"
    >
      <header>
        <p>
          foo
        </p>
        <p>
          Pod crashed!
        </p>
      </header>
      <section>
        <code>
          <span
            className={null}
            style={null}
          >
            Eeeeek the container crashed
          </span>
        </code>
      </section>
    </li>
  </ul>
</section>
`;
